bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java

```diff
@@ -0,0 +1,273 @@
+    private final Logger logger = LoggerFactory.getLogger(WebthingDiscoveryService.class);
+    private final DescriptionLoader descriptionLoader = new DescriptionLoader();
+    private final MDNSClient mdnsClient;
+
+    /**
+     * constructor
+     *
+     * @param configProperties the config props
+     * @param mdnsClient the underlying mDNS client
+     */
+    @Activate
+    public WebthingDiscoveryService(@Nullable Map<String, Object> configProperties, @Reference MDNSClient mdnsClient) {
+        super(30);
+        this.mdnsClient = mdnsClient;
+
+        super.activate(configProperties);
+        if (isBackgroundDiscoveryEnabled()) {
+            mdnsClient.addServiceListener(MDNS_SERVICE_TYPE, this);
+        }
+    }
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypes() {
+        return Set.of(THING_TYPE_UID);
+    }
+
+    @Deactivate
+    @Override
+    protected void deactivate() {
+        super.deactivate();
+        mdnsClient.removeServiceListener(MDNS_SERVICE_TYPE, this);
+    }
+
+    @Override
+    public void serviceAdded(@NonNullByDefault({}) ServiceEvent serviceEvent) {
+        considerService(serviceEvent);
+    }
+
+    @Override
+    public void serviceResolved(@NonNullByDefault({}) ServiceEvent serviceEvent) {
+        considerService(serviceEvent);
+    }
+
+    @Override
+    public void serviceRemoved(@NonNullByDefault({}) ServiceEvent serviceEvent) {
+        for (var discoveryResult : discoverWebThing(serviceEvent.getInfo())) {
+            thingRemoved(discoveryResult.getThingUID());
+        }
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        mdnsClient.addServiceListener(MDNS_SERVICE_TYPE, this);
+        startScan(true);
+    }
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+        mdnsClient.removeServiceListener(MDNS_SERVICE_TYPE, this);
+    }
+
+    private void startScan(boolean isBackground) {
+        scheduler.schedule(() -> scan(isBackground), 0, TimeUnit.SECONDS);
+    }
+
+    @Override
+    protected void startScan() {
+        startScan(false);
+    }
+
+    @Override
+    protected synchronized void stopScan() {
+        removeOlderResults(Instant.now().minus(Duration.ofMinutes(10)).toEpochMilli());
+        super.stopScan();
+    }
+
+    /**
+     * scans the network via mDNS
+     *
+     * @param isBackground true, if is background task
+     */
+    private void scan(boolean isBackground) {
+        var serviceInfos = isBackground ? mdnsClient.list(MDNS_SERVICE_TYPE)
+                : mdnsClient.list(MDNS_SERVICE_TYPE, FOREGROUND_SCAN_TIMEOUT);
+        logger.debug("got {} mDNS entries", serviceInfos.length);
+
+        // create discovery task for each detected service and process these in parallel to increase total discovery
+        // speed
+        var discoveryTasks = Arrays.stream(serviceInfos).map(DiscoveryTask::new).collect(Collectors.toList());
+        ExecutorService exec = Executors.newCachedThreadPool();
```

Comment:
You could use the ExecutorService provided by this class (`scheduler`).

Reply:
done